eth/db/storage.py

```diff
@@ -142,11 +163,46 @@ def commit_to(self, db: DatabaseAPI) -> None:
         if self._trie_nodes_batch is None:
             raise ValidationError(
                 "It is invalid to commit an account's storage if it has no pending changes. "
-                "Always check storage_lookup.has_changed_root before attempting to commit."
+                "Always check storage_lookup.has_changed_root before attempting to commit. "
+                f"Write tries on stack = {len(self._historical_write_tries)}; Root hash = "
+                f"{encode_hex(self._starting_root_hash)}"
             )
         self._trie_nodes_batch.commit_to(db, apply_deletes=False)
         self._clear_changed_root()
 
+    def new_trie(self) -> int:
+        write_trie = self._get_write_trie()
+
+        self._historical_write_tries.append((
+            write_trie,
+            self._trie_nodes_batch,
+            self._starting_root_hash,
+        ))
+
+        new_idx = len(self._historical_write_tries)
+        self._starting_root_hash = BLANK_ROOT_HASH
+        self._write_trie = None
+        self._trie_nodes_batch = None
+
+        return new_idx
+
+    def rollback_trie(self, trie_index: int) -> None:
+
+        if trie_index >= len(self._historical_write_tries):
+            raise ValidationError(
+                f"Trying to roll back a delete to index {trie_index}, but there are only"
+                f" {len(self._historical_write_tries)} indices available."
+            )
+
+        (
+            self._write_trie,
+            self._trie_nodes_batch,
+            self._starting_root_hash,
+        ) = self._historical_write_tries[trie_index]
+
+        # Cannot roll forward after a rollback, so remove created/ignored tries
+        del self._historical_write_tries[trie_index:]
```

Comment:
This also deletes the index that has just been rolled back to; it won't be possible to roll back to the same place again.

Is this intended?..

Reply:
Yup! It re-creates the index it again if there's a new account delete. I'll add a comment so it's clear that the delete is intentional.